File: kth_smallest_in_bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ans = -1;
    public int helper(TreeNode root, int k){
        if(root == null) return 0;
        int left = 1 + helper(root.left, k);
        if(left == k) ans = root.val;
        return left + helper(root.right, k - left);
    }
    public int kthSmallest(TreeNode root, int k) {
        helper(root, k);
        return ans;
    }
}

File: kth_smallest_in_bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void helper(TreeNode* root, int k, int &count, int &ans){
        if(root == nullptr) return;
        helper(root->left, k, count, ans);
        count++;
        if(k == count){
            ans = root->val;
            return;
        }
        helper(root->right, k, count, ans);
    }
    int kthSmallest(TreeNode* root, int k) {
        int ans;
        int count = 0;
        helper(root, k, count, ans);
        return ans;
    }
};